src/protocol/oprf/mod.rs

```diff
@@ -0,0 +1,330 @@
+}
+
+fn add_single_shares<'i, T, L, R>(l: L, r: R) -> impl Iterator<Item = T::Output> + 'i
+where
+    T: Add,
+    L: IntoIterator<Item = T> + 'i,
+    R: IntoIterator<Item = T> + 'i,
+{
+    l.into_iter().zip(r).map(|(a, b)| a + b)
+}
+
+// --------------------------------------------------------------------------- //
+
+fn generate_random_tables_with_peers<C: Context>(
+    batch_size: u32,
+    narrow_ctx: &C,
+) -> (
+    impl Iterator<Item = OPRFShare> + '_,
+    impl Iterator<Item = OPRFShare> + '_,
+) {
+    let (rng_l, rng_r) = narrow_ctx.prss_rng();
+    let with_left = sample_iter(rng_l).take(batch_size as usize);
+    let with_right = sample_iter(rng_r).take(batch_size as usize);
+    (with_left, with_right)
+}
+
+fn generate_random_table_solo<C>(
+    batch_size: u32,
+    narrow_ctx: &C,
+    peer: Direction,
+) -> impl Iterator<Item = OPRFShare> + '_
+where
+    C: Context,
+{
+    let rngs = narrow_ctx.prss_rng();
+    let rng = match peer {
+        Direction::Left => rngs.0,
+        Direction::Right => rngs.1,
+    };
+
+    sample_iter(rng).take(batch_size as usize)
+}
+
+fn sample_iter<R: Rng>(rng: R) -> impl Iterator<Item = OPRFShare> {
+    rng.sample_iter(Standard)
+}
+
+// ---------------------------- helper communication ------------------------------------ //
+
+async fn send_to_peer<C: Context, I: IntoIterator<Item = OPRFShare>>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    items: I,
+) -> Result<(), Error> {
+    let role = ctx.role().peer(direction);
+    let send_channel = ctx.narrow(step).send_channel(role);
+    for (record_id, row) in items.into_iter().enumerate() {
+        send_channel.send(RecordId::from(record_id), row).await?;
+    }
+    Ok(())
+}
+
+async fn receive_from_peer<C: Context>(
+    ctx: &C,
+    step: &OPRFShuffleStep,
+    direction: Direction,
+    batch_size: u32,
+) -> Result<Vec<OPRFShare>, Error> {
+    let role = ctx.role().peer(direction);
+    let receive_channel: ReceivingEnd<OPRFShare> = ctx.narrow(step).recv_channel(role);
+
+    let mut output: Vec<OPRFShare> = Vec::with_capacity(batch_size as usize);
+    for record_id in 0..batch_size {
+        let msg = receive_channel.receive(RecordId::from(record_id)).await?;
+        output.push(msg);
+    }
+
+    Ok(output)
+}
+
+// ------------------ Pseudorandom permutations functions -------------------- //
+
+fn generate_permutations_with_peers<C: Context>(batch_size: u32, ctx: &C) -> (Vec<u32>, Vec<u32>) {
+    let narrow_context = ctx.narrow(&OPRFShuffleStep::GeneratePi);
+    let mut rng = narrow_context.prss_rng();
+
+    let with_left = generate_pseudorandom_permutation(batch_size, &mut rng.0);
+    let with_right = generate_pseudorandom_permutation(batch_size, &mut rng.1);
+    (with_left, with_right)
+}
+
+fn generate_pseudorandom_permutation<R: Rng>(batch_size: u32, rng: &mut R) -> Vec<u32> {
+    let mut permutation = (0..batch_size).collect::<Vec<_>>();
+    permutation.shuffle(rng);
```

Comment:
Do we use the same function: .shuffle in our sorting protocol for generating random permutations from a seed? It seems to me that this could be generated on the fly rather than storing this huge decompressed permutation in memory.